File: cortex/enterprise/reporting.py
from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Literal

from pydantic import BaseModel, Field

from cortex.doctor import DoctorCheck, DoctorReport, run_doctor
from cortex.enterprise.config import discover_enterprise_config_path, load_enterprise_config
from cortex.enterprise.knowledge_promotion import KnowledgePromotionService

ReportingScope = Literal["local", "enterprise", "all"]


def _utc_now() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


class PromotionEventSummary(BaseModel):
    origin_id: str
    status: str
    actor: str | None = None
    updated_at: str | None = None


class PromotionReport(BaseModel):
    enabled: bool = False
    require_review: bool = True
    records_path: str | None = None
    candidates_discovered: int = 0
    candidates_ready_to_promote: int = 0
    latest_events: list[PromotionEventSummary] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class MemorySourceReport(BaseModel):
    scope: ReportingScope
    vault_path: str | None = None
    markdown_files: int = 0
    validation_errors: int = 0
    validation_warnings: int = 0
    notes: list[str] = Field(default_factory=list)


class MemoryReportPayload(BaseModel):
    generated_at: str = Field(default_factory=_utc_now)
    project_root: str
    enterprise_enabled: bool = False
    sources: list[MemorySourceReport] = Field(default_factory=list)
    promotion: PromotionReport = Field(default_factory=PromotionReport)
    doctor: dict[str, Any] = Field(default_factory=dict)


class EnterpriseReportingService:
    def __init__(self, project_root: Path) -> None:
        self.project_root = project_root.resolve()

    @staticmethod
    def from_project_root(project_root: Path) -> "EnterpriseReportingService":
        return EnterpriseReportingService(project_root)

    def build_memory_report(self, *, scope: ReportingScope = "all") -> MemoryReportPayload:
        # Always compute doctor once; we will slice it into the report.
        doctor_scope = "enterprise" if scope in {"enterprise", "all"} else "project"
        doctor = run_doctor(self.project_root, scope=doctor_scope)  # type: ignore[arg-type]

        payload = MemoryReportPayload(
            project_root=str(self.project_root),
            enterprise_enabled=discover_enterprise_config_path(self.project_root) is not None,
            doctor=_doctor_to_payload(doctor),
        )

        if scope in {"local", "all"}:
            payload.sources.append(self._local_source(doctor))

        if scope in {"enterprise", "all"}:
            enterprise = self._enterprise_source(doctor)
            if enterprise is not None:
                payload.sources.append(enterprise)

        # Promotion is enterprise-only, but we still expose a stable shape.
        if scope in {"enterprise", "all"}:
            payload.promotion = self._promotion_report()

        return payload

    def _local_source(self, doctor: DoctorReport) -> MemorySourceReport:
        vault_path = (self.project_root / "vault").resolve()
        md_count = _count_markdown_files(vault_path) if vault_path.exists() else 0
        errors = _extract_check_count(doctor.checks, "vault_validation_errors")
        warnings = _extract_check_count(doctor.checks, "vault_validation_warnings")
        notes: list[str] = []
        if not vault_path.exists():
            notes.append("vault/ directory missing")
        return MemorySourceReport(
            scope="local",
            vault_path=str(vault_path),
            markdown_files=md_count,
            validation_errors=errors,
            validation_warnings=warnings,
            notes=notes,
        )

    def _enterprise_source(self, doctor: DoctorReport) -> MemorySourceReport | None:
        cfg_path = discover_enterprise_config_path(self.project_root)
        if cfg_path is None:
            return None
        cfg = load_enterprise_config(self.project_root, required=True, path=cfg_path)
        assert cfg is not None
        enterprise_vault = cfg.resolve_enterprise_vault_path(self.project_root)
        md_count = _count_markdown_files(enterprise_vault) if enterprise_vault and enterprise_vault.exists() else 0
        errors = _extract_check_count(doctor.checks, "enterprise_vault_validation_errors")
        warnings = _extract_check_count(doctor.checks, "enterprise_vault_validation_warnings")
        notes: list[str] = []
        if enterprise_vault is None:
            notes.append("enterprise vault disabled (memory.enterprise_semantic_enabled=false)")
        elif not enterprise_vault.exists():
            notes.append("enterprise vault directory missing")
        return MemorySourceReport(
            scope="enterprise",
            vault_path=str(enterprise_vault) if enterprise_vault is not None else None,
            markdown_files=md_count,
            validation_errors=errors,
            validation_warnings=warnings,
            notes=notes,
        )

    def _promotion_report(self) -> PromotionReport:
        cfg_path = discover_enterprise_config_path(self.project_root)
        if cfg_path is None:
            return PromotionReport(enabled=False, warnings=["enterprise config missing (.cortex/org.yaml)"])
        cfg = load_enterprise_config(self.project_root, required=True, path=cfg_path)
        assert cfg is not None
        if not cfg.promotion.enabled:
            return PromotionReport(enabled=False, require_review=cfg.promotion.require_review)

        try:
            service = KnowledgePromotionService.from_project_root(self.project_root)
        except Exception as exc:
            return PromotionReport(
                enabled=True,
                require_review=cfg.promotion.require_review,
                warnings=[f"promotion reporting unavailable: {exc}"],
            )

        candidates = service.discover_candidates()
        ready = service.plan_promotion()
        latest = service.repo.load_latest_by_origin_id()
        latest_events = _summarize_latest_events(latest, limit=10)
        return PromotionReport(
            enabled=True,
            require_review=service.require_review,
            records_path=str(service.paths.records_path),
            candidates_discovered=len(candidates),
            candidates_ready_to_promote=len(ready),
            latest_events=latest_events,
        )


def _doctor_to_payload(report: DoctorReport) -> dict[str, Any]:
    return {
        "project_root": str(report.project_root),
        "checks": [
            {
                "name": c.name,
                "ok": c.ok,
                "severity": c.severity,
                "detail": c.detail,
            }
            for c in report.checks
        ],
        "has_failures": report.has_failures,
        "has_warnings": report.has_warnings,
    }


def _extract_check_count(checks: list[DoctorCheck], name: str) -> int:
    for c in checks:
        if c.name != name:
            continue
        # doctor detail is like: "<n> error(s) across ..." for these checks
        parts = str(c.detail).strip().split(" ", 1)
        if parts and parts[0].isdigit():
            return int(parts[0])
        return 0 if c.ok else 1
    return 0


def _count_markdown_files(root: Path | None) -> int:
    if root is None or not root.exists():
        return 0
    return sum(1 for _ in root.rglob("*.md"))


def _summarize_latest_events(latest_by_origin: dict[str, Any], *, limit: int) -> list[PromotionEventSummary]:
    # latest_by_origin contains PromotionRecord objects; we avoid importing models here.
    items: list[PromotionEventSummary] = []
    for origin_id, rec in latest_by_origin.items():
        status = getattr(rec, "status", "unknown")
        actor = None
        updated_at = None
        events = getattr(rec, "events", []) or []
        if events:
            last = events[-1]
            actor = getattr(last, "actor", None)
            updated_at = getattr(last, "at", None)
        items.append(PromotionEventSummary(origin_id=str(origin_id), status=str(status), actor=actor, updated_at=updated_at))
    # Sort by updated_at when present, else by origin_id.
    items.sort(key=lambda e: (e.updated_at or "", e.origin_id), reverse=True)
    return items[:limit]

File: rust/crates/cortex-enterprise/src/reporting.rs
//! Puerto de `cortex.enterprise.reporting`: reporte de memoria con doctor
//! ejecutado UNA vez vía backend inyectado.
//!
//! Seam enterprise→doctor (diseño aprobado P12B-3): enterprise define las
//! vistas neutrales `DoctorReportView`/`DoctorCheckView` y el trait
//! `DoctorBackend`. El default `UnavailableDoctorBackend` falla explícito;
//! en P12B-4 cortex-doctor implementará `NativeDoctorBackend` y el ciclo
//! queda doctor → enterprise, nunca inverso.

use std::path::PathBuf;

use serde::{Deserialize, Serialize};

use crate::config::{discover_enterprise_config_path, load_enterprise_config};
use crate::error::EnterpriseError;
use crate::knowledge_promotion::KnowledgePromotionService;
use crate::models::EnterpriseOrgConfig;
use cortex_workspace::WorkspaceLayout;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DoctorScope {
    Project,
    Enterprise,
}

/// Vistas neutrales (equivalen a DoctorCheck/DoctorReport de cortex.doctor).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DoctorCheckView {
    pub name: String,
    pub ok: bool,
    pub severity: String,
    pub detail: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DoctorReportView {
    pub project_root: PathBuf,
    pub checks: Vec<DoctorCheckView>,
    pub has_failures: bool,
    pub has_warnings: bool,
}

pub trait DoctorBackend: Send + Sync {
    fn run(
        &self,
        project_root: &std::path::Path,
        scope: DoctorScope,
    ) -> Result<DoctorReportView, EnterpriseError>;
}

/// Backend por defecto hasta P12B-4: fallo explícito contractual.
pub struct UnavailableDoctorBackend;

impl DoctorBackend for UnavailableDoctorBackend {
    fn run(
        &self,
        _project_root: &std::path::Path,
        _scope: DoctorScope,
    ) -> Result<DoctorReportView, EnterpriseError> {
        Err(EnterpriseError::BackendUnavailable(
            "doctor backend unavailable until P12B-4",
        ))
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ReportingScope {
    #[serde(rename = "local")]
    Local,
    #[serde(rename = "enterprise")]
    Enterprise,
    #[serde(rename = "all")]
    All,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PromotionEventSummary {
    pub origin_id: String,
    pub status: String,
    pub actor: Option<String>,
    pub updated_at: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PromotionReport {
    pub enabled: bool,
    pub require_review: bool,
    pub records_path: Option<String>,
    pub candidates_discovered: usize,
    pub candidates_ready_to_promote: usize,
    pub latest_events: Vec<PromotionEventSummary>,
    pub warnings: Vec<String>,
}

impl Default for PromotionReport {
    fn default() -> Self {
        Self {
            enabled: false,
            require_review: true,
            records_path: None,
            candidates_discovered: 0,
            candidates_ready_to_promote: 0,
            latest_events: vec![],
            warnings: vec![],
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemorySourceReport {
    pub scope: ReportingScope,
    pub vault_path: Option<String>,
    pub markdown_files: usize,
    pub validation_errors: usize,
    pub validation_warnings: usize,
    pub notes: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MemoryReportPayload {
    pub generated_at: String,
    pub project_root: String,
    pub enterprise_enabled: bool,
    pub sources: Vec<MemorySourceReport>,
    pub promotion: PromotionReport,
    pub doctor: serde_json::Value,
}

pub struct EnterpriseReportingService {
    project_root: PathBuf,
    layout: WorkspaceLayout,
    doctor: Box<dyn DoctorBackend>,
    clock: std::sync::Arc<dyn crate::clock::Clock>,
}

impl EnterpriseReportingService {
    /// Constructor por defecto: layout discovery + backend no disponible.
    pub fn from_project_root(
        project_root: &std::path::Path,
        layout: Option<WorkspaceLayout>,
    ) -> Result<Self, EnterpriseError> {
        let root = crate::review_knowledge::python_resolve(project_root);
        let layout = layout.unwrap_or_else(|| WorkspaceLayout::discover(&root));
        Ok(Self {
            project_root: root,
            layout,
            doctor: Box::new(UnavailableDoctorBackend),
            clock: std::sync::Arc::new(crate::clock::SystemClock),
        })
    }

    /// Inyecta el backend doctor (builder).
    pub fn with_doctor_backend(mut self, backend: impl DoctorBackend + 'static) -> Self {
        self.doctor = Box::new(backend);
        self
    }

    /// Inyecta el reloj para `generated_at` determinista (tests/gate).
    pub fn with_clock(mut self, clock: std::sync::Arc<dyn crate::clock::Clock>) -> Self {
        self.clock = clock;
        self
    }

    /// `build_memory_report`: doctor UNA vez; local→project,
    /// enterprise/all→enterprise. Las fuentes consumen el reporte por
    /// parámetro (igual que Python pasa `doctor` a `_local_source`).
    pub fn build_memory_report(
        &self,
        scope: ReportingScope,
    ) -> Result<MemoryReportPayload, EnterpriseError> {
        let doctor_scope = match scope {
            ReportingScope::Local => DoctorScope::Project,
            ReportingScope::Enterprise | ReportingScope::All => DoctorScope::Enterprise,
        };
        let doctor_report = self.doctor.run(&self.project_root, doctor_scope)?;
        let doctor_payload = doctor_to_payload(&doctor_report);

        let mut payload = MemoryReportPayload {
            generated_at: crate::clock::isoformat_seconds(self.clock.now()),
            project_root: self.project_root.display().to_string(),
            enterprise_enabled: discover_enterprise_config_path(
                &self.project_root,
                Some(&self.layout),
            )
            .is_some(),
            sources: Vec::new(),
            promotion: PromotionReport::default(),
            doctor: doctor_payload,
        };

        if matches!(scope, ReportingScope::Local | ReportingScope::All) {
            payload.sources.push(self.local_source(&doctor_report)?);
        }
        if matches!(scope, ReportingScope::Enterprise | ReportingScope::All) {
            if let Some(enterprise) = self.enterprise_source(&doctor_report)? {
                payload.sources.push(enterprise);
            }
            payload.promotion = self.promotion_report()?;
        }

        Ok(payload)
    }

    fn local_source(
        &self,
        doctor: &DoctorReportView,
    ) -> Result<MemorySourceReport, EnterpriseError> {
        let vault_path = self.layout.vault_path();
        let md_count = count_markdown_files(Some(&vault_path));
        let errors = extract_check_count(&doctor.checks, "vault_validation_errors");
        let warnings = extract_check_count(&doctor.checks, "vault_validation_warnings");
        let mut notes = Vec::new();
        if !vault_path.exists() {
            notes.push("vault directory missing".to_string());
        }
        Ok(MemorySourceReport {
            scope: ReportingScope::Local,
            vault_path: Some(vault_path.display().to_string()),
            markdown_files: md_count,
            validation_errors: errors,
            validation_warnings: warnings,
            notes,
        })
    }

    fn enterprise_source(
        &self,
        doctor: &DoctorReportView,
    ) -> Result<Option<MemorySourceReport>, EnterpriseError> {
        let Some(_cfg_path) =
            discover_enterprise_config_path(&self.project_root, Some(&self.layout))
        else {
            return Ok(None);
        };
        let cfg: EnterpriseOrgConfig =
            load_enterprise_config(&self.project_root, true, None, Some(&self.layout))?
                .ok_or_else(|| EnterpriseError::NotFound("org.yaml".to_string()))?;
        let workspace_root = self.layout.workspace_root.clone();
        let enterprise_vault =
            cfg.resolve_enterprise_vault_path(&self.project_root, Some(&workspace_root));
        let md_count = count_markdown_files(enterprise_vault.as_deref());
        let errors = extract_check_count(&doctor.checks, "enterprise_vault_validation_errors");
        let warnings = extract_check_count(&doctor.checks, "enterprise_vault_validation_warnings");
        let mut notes = Vec::new();
        if enterprise_vault.is_none() {
            notes.push(
                "enterprise vault disabled (memory.enterprise_semantic_enabled=false)".to_string(),
            );
        } else if !enterprise_vault.as_deref().is_some_and(|p| p.exists()) {
            notes.push("enterprise vault directory missing".to_string());
        }
        Ok(Some(MemorySourceReport {
            scope: ReportingScope::Enterprise,
            vault_path: enterprise_vault.map(|p| p.display().to_string()),
            markdown_files: md_count,
            validation_errors: errors,
            validation_warnings: warnings,
            notes,
        }))
    }

    fn promotion_report(&self) -> Result<PromotionReport, EnterpriseError> {
        let Some(_) = discover_enterprise_config_path(&self.project_root, Some(&self.layout))
        else {
            return Ok(PromotionReport {
                enabled: false,
                warnings: vec!["enterprise config missing (.cortex/org.yaml)".to_string()],
                ..PromotionReport::default()
            });
        };
        let cfg: EnterpriseOrgConfig =
            load_enterprise_config(&self.project_root, true, None, Some(&self.layout))?
                .ok_or_else(|| EnterpriseError::NotFound("org.yaml".to_string()))?;
        if !cfg.promotion.enabled {
            return Ok(PromotionReport {
                enabled: false,
                require_review: cfg.promotion.require_review,
                ..PromotionReport::default()
            });
        }

        let mut svc = match KnowledgePromotionService::from_project_root(
            &self.project_root,
            Some(self.layout.clone()),
            self.clock.clone(),
        ) {
            Ok(svc) => svc,
            Err(exc) => {
                return Ok(PromotionReport {
                    enabled: true,
                    require_review: cfg.promotion.require_review,
                    warnings: vec![format!("promotion reporting unavailable: {exc}")],
                    ..PromotionReport::default()
                });
            }
        };

        let candidates = svc.discover_candidates()?;
        let ready = svc.plan_promotion()?;
        let latest = svc.repo.load_latest_by_origin_id()?;
        let latest_events = summarize_latest_events(&latest, 10);
        Ok(PromotionReport {
            enabled: true,
            require_review: svc.require_review,
            records_path: Some(svc.paths.records_path.display().to_string()),
            candidates_discovered: candidates.len(),
            candidates_ready_to_promote: ready.len(),
            latest_events,
            warnings: vec![],
        })
    }
}

/// `_doctor_to_payload`.
fn doctor_to_payload(report: &DoctorReportView) -> serde_json::Value {
    serde_json::json!({
        "project_root": report.project_root.display().to_string(),
        "checks": report.checks.iter().map(|c| serde_json::json!({
            "name": c.name,
            "ok": c.ok,
            "severity": c.severity,
            "detail": c.detail,
        })).collect::<Vec<_>>(),
        "has_failures": report.has_failures,
        "has_warnings": report.has_warnings,
    })
}

/// `_extract_check_count`: parsea "<n> error(s)..." del detail; 0 si ok.
fn extract_check_count(checks: &[DoctorCheckView], name: &str) -> usize {
    for check in checks {
        if check.name != name {
            continue;
        }
        let first = check.detail.trim().split(' ').next().unwrap_or("");
        return match first.parse::<usize>() {
            Ok(n) => n,
            Err(_) => {
                if check.ok {
                    0
                } else {
                    1
                }
            }
        };
    }
    0
}

fn count_markdown_files(root: Option<&std::path::Path>) -> usize {
    let Some(root) = root else { return 0 };
    if !root.exists() {
        return 0;
    }
    let mut files = Vec::new();
    collect_md(root, &mut files);
    files.len()
}

/// `_summarize_latest_events`: sort desc por (updated_at, origin_id), top N.
fn summarize_latest_events(
    latest_by_origin: &[(String, crate::promotion_models::PromotionRecord)],
    limit: usize,
) -> Vec<PromotionEventSummary> {
    let mut items: Vec<PromotionEventSummary> = latest_by_origin
        .iter()
        .map(|(origin_id, rec)| {
            let last = rec.events.last();
            PromotionEventSummary {
                origin_id: origin_id.clone(),
                status: rec.status.as_str().to_string(),
                actor: last.and_then(|e| e.actor.clone()),
                updated_at: last.map(|e| e.at.clone()),
            }
        })
        .collect();
    items.sort_by(|a, b| {
        let ka = (
            a.updated_at.clone().unwrap_or_default(),
            a.origin_id.clone(),
        );
        let kb = (
            b.updated_at.clone().unwrap_or_default(),
            b.origin_id.clone(),
        );
        kb.cmp(&ka)
    });
    items.truncate(limit);
    items
}

fn collect_md(dir: &std::path::Path, out: &mut Vec<PathBuf>) {
    let Ok(entries) = std::fs::read_dir(dir) else {
        return;
    };
    for entry in entries.flatten() {
        let path = entry.path();
        if path.is_dir() {
            collect_md(&path, out);
        } else if path.extension().and_then(|e| e.to_str()) == Some("md") {
            out.push(path);
        }
    }
}
